src/Passes/Source/Rules/Operands/Instruction.cpp

```diff
@@ -0,0 +1,50 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+#include "Rules/OperandPrototype.hpp"
+#include "Rules/Operands/Instruction.hpp"
+
+namespace microsoft
+{
+namespace quantum
+{
+
+    template <typename T> InstructionPattern<T>::~InstructionPattern() = default;
+    template <typename T> bool InstructionPattern<T>::match(Value* instr, Captures& captures) const
+    {
+        auto* load_instr = llvm::dyn_cast<T>(instr);
+        if (load_instr == nullptr)
+        {
+            return fail(instr, captures);
+        }
+
+        return success(instr, captures);
+    }
+
+    template <typename T> typename InstructionPattern<T>::Child InstructionPattern<T>::copy() const
+    {
+        auto ret = std::make_shared<InstructionPattern<T>>();
+        ret->copyPropertiesFrom(*this);
+        return std::move(ret);
+    }
+
+// TODO(tfr): This seems to be a bug in LLVM. Template instantiations in
+// a single translation unit is not supposed to reinstantiate across other
+// translation units.
+//
+// However, it is suspecious that htis problem has been around since Clang 8.
```

Comment:
The links to the bug reports are very helpful.

Reply:
Yes, but I need to investigate this more as the error does not make any sense.